Created Package Release Guide (#8)
uasal/uasal_development_guide: source/git/git_package_release_guide.rst

Releasing a Package on Github
=============================

This guide details the pre release checks to undertake before releasing a package on github, and the actual release process on github. The guide requires no prior knowledge of creating software releases on Github. While this guide is developed for Github, the ideas themselves can be applied to releases on other git management systems (such as Gitlab).

> Preliminaries

Reviewer: Preamble?

~~~~~~~~~~~~~

The release version of the package is created by creating a stable self-contained version of the package (called a *Release*), tagging it with a version and creating accompanying metadata such as Release Notes and Changelogs. Depending on the development environment, Changelogs and version tags can be automated. Release notes are targeted at the end user, highlighting the features of the new release, whereas a Changelog is usually a detailed description of all changes since the last release version.

> The standard convention for naming versions (also known as `Semantic Versioning <https://semver.org/>`__) is to tag a release as MAJOR.MINOR.PATCH where

Reviewer: ```suggestion
The standard convention for naming versions (also known as `Semantic Versioning <https://semver.org/>`__) is to tag a release as MAJOR.MINOR.PATCH where:
```

1. MAJOR refers to an update to the whole package that is backwards incompatible with the previous versions.
2. MINOR refers to either new features, functional updates, or a large number of bundled bug fixes.

> 3. PATCH refers to small bug fixes, hotfixes, or minor changes in the code. 

Reviewer: I imagine we'll also release alphas.

Author: Added alpha/beta description

> On github, creating a release usually entails creating a stable version of the package on a branch, then creating a release tag for the specific stable commit. These branches can be standard main/develop or a dedicated release branch, which is usually dependent on the conventions of the organization creating the package. 

Reviewer: let's chat about this.

Author: We should be tagging on main for a release. Push all changes into develop, merge into main, tag the main commit as the release version. Add that "Creating a tag for release should be on main", if it is an "alpha" release, that could potentially be on develop, if features are being trialed/developed. "0.something" means development/unstable versions, once version goes to "1.something" then we don't roll over the "1." until backwards compatibility is broken. Then pushing into main should roll over the second digit, i.e. "1.1 to 1.2" etc...

Reviewer 2: Should we also change the default branch to main for our pip installs?

Author: Uhhh, I'm not sure? My best guess would be to change to main after a package is released, stay on develop while it is in active development. But also, just to clarify, is this something we need to add to the guide @AdamSchilperoort ? Seems like just a general question for package development.

Reviewer 3: I was going to bring this up actually / related to a package I was looking at and noticed we are specifying the develop branch for all of them. Maybe its best for the pip install / package end for to have no branch specified (whatever the default branch is for the repo will be used)? 

I think we don't have to change it for now though potentially (not a rush at least). Its always possible to have a branch on the tool repositories for testing out other dependencies branches for repos its using which is why I throw out the idea of having the repo dependencies to not call out to branches specifically.  

(Side Note- I believe this was just a comment / general question and didn't need to be added to the guide but wanted to add onto the conversation with Adam bringing it up)

Reviewer: +1 for default branch. Then whenever a package is stable and/or a dependency the default should become main.

Author: Sounds good, confirming that nothing needs to change on the guide!

Pre-Release Checks
~~~~~~~~~~~~~~~~~~

Prior to release, consider the following checks on your github repo:

1. Is the package on a stable release branch? Or will the package release be on the main/develop branch of the repo?
2. Does the release branch have any conflicts with the main/develop branches?
3. Do all the tests pass?

> 4. Does the release version have any dependencies/links to local files that are not in the repo?

Reviewer: I think another thing to add is that the installation instructions are detailed, with the pre-requisites specificed in the pyproject.toml file.

5. Does the code have any TODO or FIXME tags that need to be addressed before the release?
6. If using docstrings to create documentation, have all the docstrings been created/updated?
7. Does the README need to be updated?
8. Do the installation instructions need to be updated?
9. If the code comes with examples/notebooks, are these updated to work with the latest version of the code?
10. Is the release public? If so, are other users allowed to fork the repo?
11. Does the release require a discussion section? This may not be necessary with minor releases, but is encouraged for major releases.

> 12. Is the released version in alpha/beta testing?

Reviewer: normally this is why an alpha is appended to the version.
Admittedly I'm not too familiar with betas and have never released one.

Reviewer 3: I think mostly for our usage (right now), alpha would be more common than a beta release. If we have more publicly available tools for editing or even just sharing to other groups/teams for feedback, (maybe kindof like the `stp_etc` case) perhaps we should have a quick reference here for people on the differences (or link to an already existing overview guide on that)?

Author: Have added an alpha/beta description after the MAJOR.MINOR.PATCH description a little ways up on the guide. I'll resolve this comment chain for now, but there will be further opportunities to comment on the updated PR

The remainder of this guide is an abbreviated reproduction of the release guide from Github, with additional comments where deemed necessary:

1. Create a new release from the sidebar on the github repo.
2. Write down the release tag, or pick one. This should be dictated by semantic versioning.
3. Select the branch to be used as the target for the release from the *Target* dropdown menu.
4. If this is an update, the *Previous Tag* option will display previous release tags. This release tag is useful for autogenerating release notes and Changelogs, if the user chooses to do so. At the very least, it will be useful in creating an initial draft that can be suitably modified. If this is a brand new release version, no previous version will be available for selection in the *Previous tag* dropdown. 
5. Proof read the release notes for clarity, consistency and to ensure sharing of necessary and important information.
6. Add any binaries to be released alongside the code. 

> 7. If this is a *Pre-release*, check the *Set as a pre-release* checkbox.

Reviewer: We should probably say when it is appropriate to use a pre-release. Seems a bit grey.

Author: If it is a pre release, it should have "alpha" on it. Tag something as an alpha (probably on dev) and share with team that needs to use it. This is a specific use case where we tag, but don't release. Probably best to strongly discourage a pre release. So, UASAL preference is to not use pre release and just use "alpha" tagging. A github release is required to trigger a Zenodo arxived DOI. It will generate two, one is a general DOI and one is that release DOI, so each release has a new DOI. Should add this to the dev guide, how releases can be linked to Zenodo arxiving.

Author: Added this.

8. If this is set to be the latest release, check the *Set as the latest release* checkbox. This will not be visible for a first release.
9. Before finalizing, make sure that if there are any Git LFS objects in the repo, they are `managed appropriately in the release <https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-git-lfs-objects-in-archives-of-your-repository>`__.
10. Publish the release!

After the Release
~~~~~~~~~~~~~~~~~

> Releases can be edited and deleted, including updating the commit for a specific release tag, if assigned incorrectly, or if hotfixes were implemented after a release was published. While this should be done sparingly, it is a good option to fix any errors/problems with the release. 

Reviewer: We shouldn't delete releases.

Reviewer 3: ^Agree as well! Though I believe as long as branch protections are setup correctly, a user won't be able to delete them / will go through a request. (Tested this out in the past at least)

When numbering versions, in rare cases, a package may need to be released before it is tested, or even when it is in active development. These should be tagged as a `v0.MINOR.PATCH<alpha>` release and it is good practice to highlight that the release is unstable or in active development. Once a package has a first stable release, the leading digit should roll over to `v1.MINOR.PATCH<alpha>`. For all subsequent releases, the first digit should only roll over once backward compatibility is broken, so subsequent tags should go as `v1.0.0`, `v1.1.0`, `v1.2.0`, etc... 

Pre-release, release branches and archiving releases
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Note that the tagging of `v0.MINOR.PATCH<alpha>` is not the same as creating a Pre-release on GitHub, as tagging a commit is not the same as creating a release. UASAL organization preference is to *NOT utilize the Pre-release* functionality in GitHub. This is primarily to simplify the release process. This may be revisited in the future, especially if software with a large user base develops.

> UASAL organization preference is to create a release from the main branch. Prior to release, all feature branches should be merged into a develop branch, and then these changes should be merged to the main branch.

Reviewer: ```suggestion
UASAL organization preference is to create a release from the main branch. Prior to release, all desired feature branches should be merged into a develop branch, and then these changes should be merged to the main branch.
```

When releasing code alongside a paper, a user can archive a copy of the GitHub code on Zenodo. Zenodo usually generates two DOIs for each archived code, one for the code as a whole, and one for the specific release tag. So, creating a release tag will create a new snapshot of the archived code on Zenodo. Thus, in future papers, authors can choose to reference the package as a whole, or specific versions of code that were used for analysis, provided they were tagged as a GitHub release to trigger the Zenodo archiving process.

Checks before releasing a package
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Prior to release, consider the following checks on your GitHub repo:

1. Do all the tests pass?
2. Does the release version have any dependencies/links to local files that are not in the repo? 
3. Are the installation instructions detailed, with the pre-requisites specified in the pyproject.toml file?

> 4. Does the code have any TODO or FIXME tags that need to be addressed before the release?

Reviewer: ```suggestion
4. Does the code have any TODO or FIXME tags that need to be addressed before the release? In general, production code should not contain either of these tags. Issues can be created to capture this work that is to be incorporated in a future release.
```

Releasing a Package on Github
=============================

This guide details the pre release checks to undertake before releasing a package on github, and the actual release process on github. The guide requires no prior knowledge of creating software releases on Github. While this guide is developed for Github, the ideas themselves can be applied to releases on other git management systems (such as Gitlab).

Preamble
~~~~~~~~

The release version of the package is created as a stable self-contained version of the package (called a *Release*), tagging it with a version and creating accompanying metadata such as Release Notes and Changelogs. Depending on the development environment, Changelogs and version tags can be automated. Release notes are targeted at the end user, highlighting the features of the new release, whereas a Changelog is usually a detailed description of all changes since the last release version.

The standard convention for naming versions (also known as `Semantic Versioning <https://semver.org/>`__) is to tag a release as `MAJOR.MINOR.PATCH<alpha/beta>` where:

> 1. `MAJOR` refers to an update to the whole package that is backwards incompatible with the previous versions.

Reviewer: ```suggestion
1. `MAJOR` refers to an update to the whole package that is not backwards compatible with the previous versions.
```

On github, creating a release usually entails creating a stable version of the package on a branch, then creating a release tag for the specific stable commit. These branches can be standard main/develop or a dedicated release branch, which is usually dependent on the conventions of the organization creating the package. 

When numbering versions, in rare cases, a package may need to be released before it is tested, or even when it is in active development. These should be tagged as a `v0.MINOR.PATCH<alpha>` release and it is good practice to highlight that the release is unstable or in active development. Once a package has a first stable release, the leading digit should roll over to `v1.MINOR.PATCH<alpha>`. For all subsequent releases, the first digit should only roll over once backward compatibility is broken, so subsequent tags should go as `v1.0.0`, `v1.1.0`, `v1.2.0`, etc... 

Pre-release, release branches and archiving releases
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Note that the tagging of `v0.MINOR.PATCH<alpha>` is not the same as creating a Pre-release on github, as tagging a commit is not the same as creating a release. UASAL organization preference is to *NOT create Pre-release versions* of packages.

Reviewer: ```suggestion
Note that the tagging of `v0.MINOR.PATCH<alpha>` is not the same as creating a Pre-release on github, as tagging a commit is not the same as creating a release. UASAL organization preference is to *NOT utilize the Pre-release* functionality in GitHub. This is primarily to simplify the release process. This may be revisited in the future, especially if software with a large user base develops.
```

> UASAL organization preference is to tag a release on the main branch. Prior to release, all feature branches should be merged into a develop branch, and then these changes should be pushed to the main branch and then tagged as a release from there. 

Reviewer: ```suggestion
UASAL organization preference is to create a release from the main branch. Prior to release, all feature branches should be merged into a develop branch, and then these changes should be merged to the main branch.
```

When releasing code alongside a paper, a user can archive a copy of the github code on Zenodo. Zenodo usually generates two DOIs for each archived code, one for the code as a whole, and one for the specific release tag. So, creating a release tag will create a new snapshot of the archived code on Zenodo. Thus, in future papers, authors can choose to reference the package as a whole, or specific versions of code that were used for analysis, provided they were tagged as a github release to trigger the Zenodo archiving process.

Checks before releasing a package
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Prior to release, consider the following checks on your github repo:

> 1. Is the package on a stable release branch? Or will the package release be on the main/develop branch of the repo?

Reviewer: I think 1 and 2 are no longer needed since we are recommending to release from main?

2. Does the release branch have any conflicts with the main/develop branches?
3. Do all the tests pass?
4. Does the release version have any dependencies/links to local files that are not in the repo? 
5. Are the installation instructions detailed, with the pre-requisites specified in the pyproject.toml file?
6. Does the code have any TODO or FIXME tags that need to be addressed before the release?
7. If using docstrings to create documentation, have all the docstrings been created/updated?

> 8. Does the README need to be updated?

Reviewer: ```suggestion
8. Does the README or other documentation need to be updated?
```

9. Do the installation instructions need to be updated?

> 10. If the code comes with examples/notebooks, are these updated to work with the latest version of the code?

Reviewer: Notebooks (in the docs section) should also be wiped of content, whereas notebooks in the notebook directory (examples) can have content.

11. Is the release public? If so, are other users allowed to fork the repo?
12. Does the release require a discussion section? This may not be necessary with minor releases, but is encouraged for major releases.
13. Is the released version in alpha/beta testing?

Release Procedure
~~~~~~~~~~~~~~~~~

Once all checks have been completed and any issues have been resolved, users can draft a new release `On Github <https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository>`__

The remainder of this guide is an abbreviated reproduction of the release guide from Github, with additional comments where deemed necessary:

> 1. Create a new release from the sidebar on the github repo.

Reviewer: Consider doing a find/replace for "github" to be "GitHub"

> 2. Write down the release tag, or pick one. This should be dictated by semantic versioning.

Reviewer: ```suggestion
2. Determine the appropriate release version. This should be dictated by semantic versioning.
```

3. Select the branch to be used as the target for the release from the *Target* dropdown menu.
4. If this is an update, the *Previous Tag* option will display previous release tags. This release tag is useful for autogenerating release notes and Changelogs, if the user chooses to do so. At the very least, it will be useful in creating an initial draft that can be suitably modified. If this is a brand new release version, no previous version will be available for selection in the *Previous tag* dropdown. 
5. Proof read the release notes for clarity, consistency and to ensure sharing of necessary and important information.
6. Add any binaries to be released alongside the code. 

> 7. If this is a *Pre-release*, check the *Set as a pre-release* checkbox.

Reviewer: I think you can delete this since we decided not to use them?

Reviewer 3: Agree with this ^
